class.jetpack.php

```diff
@@ -1905,6 +1905,9 @@ function extra_oembed_providers() {
 		wp_oembed_add_provider( '#https?://[^.]+\.(wistia\.com|wi\.st)/(medias|embed)/.*#', 'https://fast.wistia.com/oembed', true );
 		wp_oembed_add_provider( '#https?://sketchfab\.com/.*#i', 'https://sketchfab.com/oembed', true );
 		wp_oembed_add_provider( '#https?://(www\.)?icloud\.com/keynote/.*#i', 'https://iwmb.icloud.com/iwmb/oembed', true );
+
+		// For core, see <https://core.trac.wordpress.org/ticket/46067>.
+		wp_oembed_add_provider( '#^https?://giphy\.com/embed/(?:.+?-)?\w+#', 'https://giphy.com/services/oembed', true );
```

Comment:
I'm curious why this block is not more heavily relying on oEmbed. It could be an embed block very similar to the others in core, and then most of the markup for the block could be static with a bare URL which is converted into the iframe. I suppose the only problem with this is the Giphy oEmbed response doesn't serve back an iframe, so we'd need to filter it as well to replace the img with an iframe. This filter is where the AMP-compat code could go as well.

Reply:
@jeffersonrabb Could you address Weston's question?

Reply:
Looking at this PR today—sorry I've let it sit so long.

Reply:
Just to make sure I understand the question, @westonruter you're suggesting that oEmbed be used universally for the block, not just for AMP pages as implemented here?

Reply:
@jeffersonrabb hey! Yes, I'm curious why oEmbed isn't being used generally for implementing Giphy support.

Reply:
No good reason! The block is an adaptation of a much older block from a different system, which doesn't use oEmbed and simply rendered the iFrame. I'm researching the oEmbed now approach and will get back shortly. Looks like this might be an easy win. If switching to oEmbed does make sense, would that means we toss this PR and start fresh? Or would it make sense for me to implement AMP compatibility in the new work?

Reply:
It's a good question. I don't think the iframe is actually bad to use because it actually provides a better experience than the oEmbed. 

Given this Giphy: https://giphy.com/gifs/mars-erosion-bJc1o7QdHRBHq

The oEmbed URL is:

https://giphy.com/services/oembed?url=https://media.giphy.com/media/bJc1o7QdHRBHq/giphy.gif

```json
{
    "author_name": "GIPHY",
    "author_url": "https://giphy.com/",
    "height": 250,
    "provider_name": "GIPHY",
    "provider_url": "https://giphy.com/",
    "title": "Mars GIF - Find & Share on GIPHY",
    "type": "photo",
    "url": "https://media.giphy.com/media/bJc1o7QdHRBHq/giphy.gif",
    "width": 250
}
```

This as you can see returns an actual GIF (`image/gif`). On the other hand, if you do the iframe embed:

```html
<iframe src="https://giphy.com/embed/bJc1o7QdHRBHq" width="480" height="480" frameBorder="0" class="giphy-embed" allowFullScreen></iframe><p><a href="https://giphy.com/gifs/mars-erosion-bJc1o7QdHRBHq">via GIPHY</a></p>
```

This loads https://giphy.com/embed/bJc1o7QdHRBHq which embeds a WebP image.

The WebP image is 121KB whereas the GIF is 334KB, and other animations would surely have a much larger difference between GIF and WebP, as GIF is notorious for huge image sizes.

The problem with Giphy's oEmbed is that they only support the `photo` type, not the `rich` type (which would allow `html` and an `iframe`).

All of this to say, I think there could be a hybrid approach. Jetpack could go ahead and go the oEmbed route but then use the `embed_oembed_html` filter to automatically upgrade the `photo` response into an `iframe`, similar to how you've implemented embedding here.

The benefit here is that a user should be able to paste the Giphy URL into a generic Embed block, and it would just work.

Reply:
This same `embed_oembed_html` filter function could output both an `<iframe>` in non-AMP and an `<amp-giphy>` in AMP (when `Jetpack_AMP_Support::is_amp_request()`).

Reply:
Thanks for all the detail! When you say "just work" do you mean the iFrame approach?

Reply:
Yes, with the above oEmbed and filter implemented, a regular Embed block could take a Giphy URL and automatically render an iframe. Still having a dedicated Giphy Embed block should be retained like you have here so that you can have a nice search interface to find the GIF to embed. But once it has been selected you can then let oEmbed do the rest.

Reply:
I like it. Are you aware if there are other WP oEmbed implementations that upgrade a simple media file response to an iFrame?

Reply:
There are examples of doing this for AMP in https://github.com/ampproject/amp-wp/tree/develop/includes/embeds

In particular, see:

* https://github.com/ampproject/amp-wp/blob/develop/includes/embeds/class-amp-imgur-embed-handler.php
* https://github.com/ampproject/amp-wp/blob/develop/includes/embeds/class-amp-gfycat-embed-handler.php

Reply:
I implemented this for Giphy in 8f339ede39fa89bbafa23e0ded4181213d55b380. This is probably not the right location for this function, but I'm sure you know where it would best live.

Reply:
So then the next step I think would be to rework the GIF block to just output the selected Giphy URL on a single line in the block. Then the oEmbed handling will automatically convert it.

Reply:
OK great, I'll run with it. I'm slightly blocked because my local environment—typical Jetpack Docker instance—500s whenever an AMP URL is loaded (e.g. `http://localhost/2019/03/gif/?amp`). Result of `yarn docker:tail` for one such failure below. I've tried using a public URL too, Any of this ring a bell?

```[16-Mar-2019 01:48:07 UTC] PHP Fatal error:  Maximum execution time of 30 seconds exceeded in /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php on line 2104
[16-Mar-2019 01:48:07 UTC] PHP Stack trace:
[16-Mar-2019 01:48:07 UTC] PHP   1. shutdown_action_hook() /var/www/html/wp-includes/load.php:0
[16-Mar-2019 01:48:07 UTC] PHP   2. do_action() /var/www/html/wp-includes/load.php:788
[16-Mar-2019 01:48:07 UTC] PHP   3. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:465
[16-Mar-2019 01:48:07 UTC] PHP   4. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[16-Mar-2019 01:48:07 UTC] PHP   5. wp_ob_end_flush_all() /var/www/html/wp-includes/class-wp-hook.php:286
[16-Mar-2019 01:48:07 UTC] PHP   6. ob_end_flush() /var/www/html/wp-includes/functions.php:4212
[16-Mar-2019 01:48:07 UTC] PHP   7. AMP_Theme_Support::finish_output_buffering() /var/www/html/wp-includes/functions.php:4212
[16-Mar-2019 01:48:07 UTC] PHP   8. AMP_Theme_Support::prepare_response() /var/www/html/wp-content/plugins/amp/includes/class-amp-theme-support.php:1473
[16-Mar-2019 01:48:07 UTC] PHP   9. AMP_Content_Sanitizer::sanitize_document() /var/www/html/wp-content/plugins/amp/includes/class-amp-theme-support.php:1727
[16-Mar-2019 01:48:07 UTC] PHP  10. AMP_Style_Sanitizer->sanitize() /var/www/html/wp-content/plugins/amp/includes/templates/class-amp-content-sanitizer.php:107
[16-Mar-2019 01:48:07 UTC] PHP  11. AMP_Style_Sanitizer->process_link_element() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:493
[16-Mar-2019 01:48:07 UTC] PHP  12. AMP_Style_Sanitizer->process_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:751
[16-Mar-2019 01:48:07 UTC] PHP  13. AMP_Style_Sanitizer->prepare_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:862
[16-Mar-2019 01:48:07 UTC] PHP  14. AMP_Style_Sanitizer->parse_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1096
[16-Mar-2019 01:48:07 UTC] PHP  15. AMP_Style_Sanitizer->process_css_list() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1033
[16-Mar-2019 01:48:07 UTC] PHP  16. AMP_Style_Sanitizer->process_css_declaration_block() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1298
[16-Mar-2019 01:48:07 UTC] PHP  17. AMP_Style_Sanitizer->ampify_ruleset_selectors() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1446
[16-Mar-2019 01:48:07 UTC] PHP Fatal error:  Unknown: Cannot use output buffering in output buffering display handlers in Unknown on line 0
[16-Mar-2019 01:48:07 UTC] PHP Stack trace:
[16-Mar-2019 01:48:07 UTC] PHP   1. shutdown_action_hook() /var/www/html/wp-includes/load.php:0
[16-Mar-2019 01:48:07 UTC] PHP   2. do_action() /var/www/html/wp-includes/load.php:788
[16-Mar-2019 01:48:07 UTC] PHP   3. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:465
[16-Mar-2019 01:48:07 UTC] PHP   4. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[16-Mar-2019 01:48:07 UTC] PHP   5. wp_ob_end_flush_all() /var/www/html/wp-includes/class-wp-hook.php:286
[16-Mar-2019 01:48:07 UTC] PHP   6. ob_end_flush() /var/www/html/wp-includes/functions.php:4212
[16-Mar-2019 01:48:07 UTC] PHP   7. AMP_Theme_Support::finish_output_buffering() /var/www/html/wp-includes/functions.php:4212
[16-Mar-2019 01:48:07 UTC] PHP   8. AMP_Theme_Support::prepare_response() /var/www/html/wp-content/plugins/amp/includes/class-amp-theme-support.php:1473
[16-Mar-2019 01:48:07 UTC] PHP   9. AMP_Content_Sanitizer::sanitize_document() /var/www/html/wp-content/plugins/amp/includes/class-amp-theme-support.php:1727
[16-Mar-2019 01:48:07 UTC] PHP  10. AMP_Style_Sanitizer->sanitize() /var/www/html/wp-content/plugins/amp/includes/templates/class-amp-content-sanitizer.php:107
[16-Mar-2019 01:48:07 UTC] PHP  11. AMP_Style_Sanitizer->process_link_element() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:493
[16-Mar-2019 01:48:07 UTC] PHP  12. AMP_Style_Sanitizer->process_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:751
[16-Mar-2019 01:48:07 UTC] PHP  13. AMP_Style_Sanitizer->prepare_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:862
[16-Mar-2019 01:48:07 UTC] PHP  14. AMP_Style_Sanitizer->parse_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1096
[16-Mar-2019 01:48:07 UTC] PHP  15. AMP_Style_Sanitizer->process_css_list() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1033
[16-Mar-2019 01:48:07 UTC] PHP  16. AMP_Style_Sanitizer->process_css_declaration_block() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1298
[16-Mar-2019 01:48:07 UTC] PHP  17. AMP_Style_Sanitizer->ampify_ruleset_selectors() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1446
[16-Mar-2019 01:50:56 UTC] SANDBOXING via 'jrjp304047161.wordpress.com': 'https://jetpack.com/psh-status/'
```

Reply:
Do you have `display_errors` turned on?

The error:

> Cannot use output buffering in output buffering display handlers in Unknown on line 0

This may be due to a warning being printed out during post processing of the output buffer. See https://github.com/ampproject/amp-wp/issues/1915 for a fix planned in the plugin.

But if that is the case, I want to know what the underlying warning was to trigger the error. Is there any previous log entry with a warning perhaps?

Reply:
There is this warning:

`WARNING: The PORT_WORDPRESS variable is not set. Defaulting to a blank string.`

Reply:
Humm. That seems like it's something specific to your environment. Where is `PORT_WORDPRESS` being read?

Also, try applying this patch to your AMP plugin:

```diff
diff --git a/includes/class-amp-theme-support.php b/includes/class-amp-theme-support.php
index 507919ec..2dd8a6b4 100644
--- a/includes/class-amp-theme-support.php
+++ b/includes/class-amp-theme-support.php
@@ -1672,6 +1672,8 @@ class AMP_Theme_Support {
 	 */
 	public static function prepare_response( $response, $args = array() ) {
 		global $content_width;
+		ini_set( 'display_errors', 0 );
+
 		$prepare_response_start = microtime( true );
 
 		if ( isset( $args['validation_error_callback'] ) ) {
```

Reply:
Otherwise, it seems there is an error happening in `\AMP_Style_Sanitizer::ampify_ruleset_selectors()` but it's not clear what it is from the stack trace.

Is also related to running out of memory as was reported by someone else in https://github.com/ampproject/amp-wp/issues/1915? So what happens if you modify your `WP_MEMORY_LIMIT` in `wp-config.php` to:

```php
define( 'WP_MEMORY_LIMIT', '128M' );
```

Reply:
`PORT_WORDPRESS` is indeed something from our Docker environment. @jeffersonrabb you can bypass this via this section of https://github.com/Automattic/jetpack/blob/master/docker/README.md

> You can set the following variables on a per-command basis (PORT_WORDPRESS=80 yarn docker:up) or, preferably, in a ./.env file in Jetpack's root directory.
> PORT_WORDPRESS: (default=80) The port on your host machine connected to the WordPress container's HTTP server.

Opening https://github.com/Automattic/jetpack/issues/11583 to track making that warning display only at more actionable times to avoid false issues like this. Sorry about that folks!